### PR TITLE
[DeviceSanitizer] Support detecting memory leaks of USM

### PR DESCRIPTION
**Pull request by zhaomaosu**
Also maintain an internal ref count of ContextInfo.

SYCLOS Part: https://github.com/intel/llvm/pull/14378

### CODE REVIEW

**Review by yingcong-wu — APPROVED** [review]
lgtm.

File: source/loader/layers/sanitizer/ur_sanddi.cpp [hunk 1]
```diff
@@ -358,6 +358,29 @@ __urdlllocal ur_result_t UR_APICALL urContextCreateWithNativeHandle(
...
+        return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    }
+
+    context.logger.debug("==== urContextRetain");
+
+    UR_CALL(pfnRetain(hContext));
+
+    if (auto ContextInfo = context.interceptor->getContextInfo(hContext)) {
+        ContextInfo->RefCount++;
+    }
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/ur_sanddi.cpp:379]: I think we can very sure `assert(ContextInfo != nullptr)`.

File: source/loader/layers/sanitizer/ur_sanddi.cpp [hunk 2]
```diff
@@ -371,10 +394,16 @@ __urdlllocal ur_result_t UR_APICALL urContextRelease(
 
     context.logger.debug("==== urContextRelease");
 
-    UR_CALL(context.interceptor->eraseContext(hContext));
-    ur_result_t result = pfnRelease(hContext);
+    UR_CALL(pfnRelease(hContext));
 
-    return result;
+    if (auto ContextInfo = context.interceptor->getContextInfo(hContext)) {
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/ur_sanddi.cpp:399]: I think we can very sure `assert(ContextInfo != nullptr)`.

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 3]
```diff
@@ -153,6 +153,13 @@ SanitizerInterceptor::SanitizerInterceptor() {
 SanitizerInterceptor::~SanitizerInterceptor() {
     DestroyShadowMemoryOnCPU();
     DestroyShadowMemoryOnPVC();
+
+    if (!m_AllocationMap.empty()) {
+        for (const auto &[_, AI] : m_AllocationMap) {
+            ReportMemoryLeak(AI);
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.cpp:159]: skip `AI->IsReleased == true`

File: source/loader/layers/sanitizer/ur_sanddi.cpp (hunk 1 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/ur_sanddi.cpp]: Done

File: source/loader/layers/sanitizer/asan_interceptor.cpp (hunk 3 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:161]: Done

Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:161]: Oh, you also need to use mutex

Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:162]: Thanks, done.

**Review by AllanZyne — APPROVED** [review]

**Comment by zhaomaosu** [comment]
Hi @oneapi-src/unified-runtime-maintain, could you please help review this PR? There is a failure in pre-ci, but I didn't see any error message, looks so weird.

**Pull request merged**